### PR TITLE
cleanup(pubsublite): have routing policy return std::uint32_t

### PR DESCRIPTION
**Pull request by 18suresha**
This PR cleans up `RoutingPolicy` so that it returns an `std::uint32_t` so it can be consistent with future code that uses partition-based logic.
It also adds a few assertions for boundary (UINT32_MAX - 1) mod values

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8675)

### CODE REVIEW

File: google/cloud/pubsublite/internal/default_routing_policy.cc [hunk 1]
```diff
@@ -33,29 +37,31 @@ std::uint64_t ModPow(std::uint64_t val, std::uint64_t pow, std::uint32_t mod) {
...
-std::uint64_t DefaultRoutingPolicy::Route(std::uint32_t num_partitions) {
+std::uint32_t DefaultRoutingPolicy::Route(Partition num_partitions) {
   // atomic operation
-  return counter_++ % num_partitions;
+  return static_cast<std::uint32_t>(counter_++ % num_partitions);
 }
 
-std::uint64_t DefaultRoutingPolicy::Route(std::string const& message_key,
-                                          std::uint32_t num_partitions) {
+std::uint32_t DefaultRoutingPolicy::Route(std::string const& message_key,
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/default_routing_policy.cc:63]: Partition

File: google/cloud/pubsublite/internal/default_routing_policy.cc [hunk 2]
```diff
@@ -33,29 +37,31 @@ std::uint64_t ModPow(std::uint64_t val, std::uint64_t pow, std::uint32_t mod) {
...
     result += (val_mod * pow_mod) % mod;
     result %= mod;
   }
-  return result;
+  // within bounds because `mod`ed by a std::uint32_t value
+  return static_cast<std::uint32_t>(result);
 }
 
-std::uint64_t DefaultRoutingPolicy::Route(std::uint32_t num_partitions) {
+std::uint32_t DefaultRoutingPolicy::Route(Partition num_partitions) {
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/default_routing_policy.cc:58]: Partition

File: google/cloud/pubsublite/internal/default_routing_policy.cc [hunk 3]
```diff
@@ -21,10 +22,13 @@ namespace cloud {
...
 
+// will always be 32 as specified in sha256_hash.h
+auto constexpr kNumBytesSha256 = 32;
+
 // Uses the identity that `(a*b) % m == ((a % m) * (b % m)) % m`
-std::uint64_t ModPow(std::uint64_t val, std::uint64_t pow, std::uint32_t mod) {
+std::uint64_t ModPow(std::uint64_t val, unsigned int pow, std::uint32_t mod) {
   std::uint64_t result = 1;
-  for (std::uint32_t i = 0; i < pow; ++i) {
+  for (unsigned int i = 0; i != pow; ++i) {
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/default_routing_policy.cc:31]: please revert here and elsewhere, don't use unsigned int

File: google/cloud/pubsublite/internal/routing_policy.h [hunk 4]
```diff
@@ -29,10 +29,11 @@ namespace pubsublite_internal {
  */
 class RoutingPolicy {
  public:
+  using Partition = std::uint32_t;
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/routing_policy.h:32]: I'd just elevate this to the header, not put it in the class body

File: google/cloud/pubsublite/internal/default_routing_policy.h [hunk 5]
```diff
@@ -30,10 +29,10 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 // Calculates the val^pow % mod while accounting for overflow.
 // Needed because after calculating `big_endian[i]` % `mod` in `GetMod`, we need
 // to account for its position in the array by multiplying it by an offset.
-std::uint64_t ModPow(std::uint64_t val, std::uint64_t pow, std::uint32_t mod);
+std::uint64_t ModPow(std::uint64_t val, std::uint64_t pow, Partition mod);
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/default_routing_policy.h:32]: nit. for "math" functions like ModPow and GetMod, probably better to just use std::uint32_t, makes it easier to think about overflows.

**Review by dpcollins-google — APPROVED** [review]

**Review by coryan — APPROVED** [review]

**Pull request merged**